ArduCopter/Attitude.cpp

```diff
@@ -94,31 +94,35 @@ float Copter::get_look_ahead_yaw()
  *  throttle control
  ****************************************************************/
 
-// update_thr_average - update estimated throttle required to hover (if necessary)
-//  should be called at 100hz
-void Copter::update_thr_average()
+// update estimated throttle required to hover (if necessary)
+//  called at 100hz
+void Copter::update_throttle_hover()
 {
-    // ensure throttle_average has been initialised
-    if( is_zero(throttle_average) ) {
-        throttle_average = 0.5f;
-        // update position controller
-        pos_control.set_throttle_hover(throttle_average);
-    }
-
+#if FRAME_CONFIG != HELI_FRAME
     // if not armed or landed exit
     if (!motors.armed() || ap.land_complete) {
         return;
     }
 
+    // do not update in manual throttle modes or Drift
+    if (mode_has_manual_throttle(control_mode) || (control_mode == DRIFT)) {
+        return;
+    }
+
+    // do not update while climbing or descending
+    if (!is_zero(pos_control.get_desired_velocity().z)) {
+        return;
+    }
```

Comment:
That's a very strict condition to meet Zspeed=0. Perhaps a small deadband around zero, optionally with a user adjustable parameter for HoverPrecision that would define the deadband?